File: core/src/main/java/com/wbrawner/twigs/permission/UserPermissionRepository.java
package com.wbrawner.twigs.permission;

import com.wbrawner.twigs.budget.Budget;
import com.wbrawner.twigs.user.User;
import org.springframework.data.domain.Pageable;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.PagingAndSortingRepository;

import java.util.List;
import java.util.Optional;

public interface UserPermissionRepository extends CrudRepository<UserPermission, UserPermissionKey>,
        PagingAndSortingRepository<UserPermission, UserPermissionKey> {
    Optional<UserPermission> findByUserAndBudget_Id(User user, String budgetId);

    List<UserPermission> findAllByUser(User user, Pageable pageable);

    List<UserPermission> findAllByBudget(Budget budget, Pageable pageable);

    List<UserPermission> findAllByUserAndBudget(User user, Budget budget, Pageable pageable);

    List<UserPermission> findAllByUserAndBudget_IdIn(User user, List<String> budgetIds, Pageable pageable);

    void deleteAllByBudget(Budget budget);
}

File: api/src/main/kotlin/com/wbrawner/twigs/server/permission/UserPermissionRepository.kt
package com.wbrawner.twigs.server.permission

import com.wbrawner.twigs.server.budget.Budget
import com.wbrawner.twigs.server.user.User
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.PagingAndSortingRepository
import java.util.*

interface UserPermissionRepository : PagingAndSortingRepository<UserPermission, UserPermissionKey> {
    fun findByUserAndBudget_Id(user: User?, budgetId: String?): Optional<UserPermission>
    fun findAllByUser(user: User?, pageable: Pageable?): List<UserPermission>
    fun findAllByBudget(budget: Budget?, pageable: Pageable?): List<UserPermission>
    fun findAllByUserAndBudget(user: User?, budget: Budget?, pageable: Pageable?): List<UserPermission>
    fun findAllByUserAndBudget_IdIn(user: User?, budgetIds: List<String?>?, pageable: Pageable?): List<UserPermission>
}